search/searcher/term.go

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package searcher
+
+import (
+	"github.com/m3db/m3ninx/index"
+	"github.com/m3db/m3ninx/postings"
+	"github.com/m3db/m3ninx/search"
+)
+
+type termSearcher struct {
+	field, term []byte
+	readers     index.Readers
+
+	idx  int
+	curr postings.List
+
+	closed bool
+	err    error
+}
+
+// NewTermSearcher returns a new searcher for finding documents which match the given term.
+// It is not safe for concurrent access.
+func NewTermSearcher(rs index.Readers, field, term []byte) search.Searcher {
```

Comment:
nit: should you take a `Snapshot` as input here? that way you don't need to clone the readers slice, can reference count on the Snapshot itself

Reply:
I think it would be simpler if we only reference counted the segments. That's why we would take a reference count on the Snapshot, we would transitively being taking a reference count on the segment.

Reply:
For sure its simpler to have the readers cloned but i'm concerned about the allocs & contention. 

For e.g. if you have a disjunction query with 3 sub-parts, you'd end up 4 different clone of the Readers. All the while you only needed to guarantee the segment's are valid for the duration of the top level query. Which a single clone of the readers is sufficient for.

Reply:
Talked offline, we won't be cloning the Readers anymore. The top-level `Executor` will be responsible for enforcing their lifetimes.